File: src/commands/update.ts
import { chmodSync, renameSync, unlinkSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import * as prompts from "@clack/prompts";
import { Command } from "commander";
import pkg from "../../package.json" with { type: "json" };
import {
	getDownloadUrl,
	getLatestRelease,
	getPlatformSlug,
	isUpdateAvailable,
} from "../utils/update";

export const update = new Command("update")
	.description("Update scli to the latest version")
	.option("--check", "Check for updates without installing")
	.action(async (opts: { check?: boolean }) => {
		prompts.intro("update");

		const spinner = prompts.spinner();
		spinner.start("Checking for latest release…");

		let latest: Awaited<ReturnType<typeof getLatestRelease>>;
		try {
			latest = await getLatestRelease();
		} catch (err) {
			spinner.stop("Failed to reach GitHub.");
			prompts.log.error(String(err));
			process.exit(1);
		}

		const current = pkg.version;
		const hasUpdate = isUpdateAvailable(current, latest.version);

		if (!hasUpdate) {
			spinner.stop(`Already up to date (v${current}).`);
			prompts.outro("Nothing to do.");
			return;
		}

		spinner.stop(
			`New version available: v${latest.version} (current: v${current})`,
		);

		if (opts.check) {
			prompts.log.info(`Run \`scli update\` to install v${latest.version}.`);
			prompts.outro("Done.");
			return;
		}

		const confirmed = await prompts.confirm({
			message: `Install v${latest.version}?`,
		});

		if (prompts.isCancel(confirmed) || !confirmed) {
			prompts.cancel("Update cancelled.");
			process.exit(0);
		}

		let slug: string;
		try {
			slug = getPlatformSlug();
		} catch (err) {
			prompts.log.error(String(err));
			process.exit(1);
		}

		const url = getDownloadUrl(latest.tag, slug);
		spinner.start(`Downloading ${url}…`);

		let response: Response;
		try {
			response = await fetch(url);
		} catch (err) {
			spinner.stop("Download failed.");
			prompts.log.error(String(err));
			process.exit(1);
		}

		if (!response.ok) {
			spinner.stop(
				`Download failed: ${response.status} ${response.statusText}`,
			);
			process.exit(1);
		}

		const buffer = await response.arrayBuffer();
		spinner.stop("Download complete.");

		const tmpPath = join(tmpdir(), `scli-update-${Date.now()}`);
		writeFileSync(tmpPath, Buffer.from(buffer));
		chmodSync(tmpPath, 0o755);

		try {
			renameSync(tmpPath, process.execPath);
		} catch {
			// rename across devices fails on some systems — fall back to copy+delete
			const dest = Bun.file(process.execPath);
			await Bun.write(dest, Bun.file(tmpPath));
			chmodSync(process.execPath, 0o755);
			try {
				unlinkSync(tmpPath);
			} catch {
				/* ignore */
			}
		}

		prompts.outro(
			`Updated to v${latest.version}. Restart scli to use the new version.`,
		);
	});

File: src/commands/update.rs
use anyhow::Result;
use dialoguer::{Confirm, theme::ColorfulTheme};
use crate::utils::update::{get_latest_release, is_update_available, get_platform_slug, get_download_url};


pub fn execute_update(check: bool) -> Result<()> {
    println!("Checking for latest release...");

    let latest = match get_latest_release() {
        Ok(r) => r,
        Err(e) => {
            eprintln!("Failed to reach GitHub.");
            eprintln!("{e}");
            std::process::exit(1);
        }
    };

    let current = env!("CARGO_PKG_VERSION");
    let has_update = is_update_available(current, &latest.version);

    if !has_update {
        println!("Already up to date (v{current}).");
        println!("Nothing to do.");
        return Ok(());
    }

    println!("New version available: v{} (current: v{current})", latest.version);

    if !latest.notes.is_empty() {
        println!("\n--- Release Notes ---\n{}\n---------------------", latest.notes);
    }

    if check {
        println!("Run `scli update` to install v{}.", latest.version);
        return Ok(());
    }

    let confirmed = Confirm::with_theme(&ColorfulTheme::default())
        .with_prompt(format!("Install v{}?", latest.version))
        .interact()?;

    if !confirmed {
        println!("Update cancelled.");
        return Ok(());
    }

    let slug = match get_platform_slug() {
        Ok(s) => s,
        Err(e) => {
            eprintln!("{e}");
            std::process::exit(1);
        }
    };

    let url = get_download_url(&latest.tag, &slug);
    println!("Downloading {url}...");

    let response = match ureq::get(&url).call() {
        Ok(r) => r,
        Err(e) => {
            eprintln!("Download failed: {e}");
            std::process::exit(1);
        }
    };

    let body = response.into_body().read_to_vec()?;
    println!("Download complete.");

    let tmp_path = std::env::temp_dir().join(format!("scli-update-{}", std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_millis()));
    std::fs::write(&tmp_path, &body)?;

    use std::os::unix::fs::PermissionsExt;
    std::fs::set_permissions(&tmp_path, std::fs::Permissions::from_mode(0o755))?;

    let exec_path = std::env::current_exe()?;

    if std::fs::rename(&tmp_path, &exec_path).is_err() {
        std::fs::copy(&tmp_path, &exec_path)?;
        std::fs::set_permissions(&exec_path, std::fs::Permissions::from_mode(0o755))?;
        let _ = std::fs::remove_file(&tmp_path);
    }

    println!("Updated to v{}. Restart scli to use the new version.", latest.version);
    Ok(())
}
